doc: add CODE_OVERVIEW.md

the-real-dill opened the pull request:
Introduces a new documentation file, `CODE_OVERVIEW.md`, at the root of the repository. The aim of this files is to provide a high-level overview of the XLibre server codebase, to assist new contributors in navigating and understanding its structure.

Includes:
- An explanation of core principles like modularity, platform independence, and build flexibility.
- An ASCII art diagram illustrating the layering of components (OS abstractions, DIX core, extensions, DDX hardware/servers).
- A directory-by-directory breakdown with summaries of purpose, key files, and interconnections to other parts of the codebase.
- Practical tips for developers, covering entry points, debugging, modules, references to additional documentation, contributing tips, and community resources.

This document serves as a starting point for new developers trying to understand the codebase, and is not exhaustive reference material. Future updates can expand on it.

I am not an expert on X11 or this codebase, this was written as I was trying to understand the codebase, and I thought it may be useful to others. If you want to merge this, it would be good to have someone with more experience to review it for errors or omissions.

I thought this fits best at the root of the codebase, instead of say the wiki, as it is an artifact related directly to the code, for people working on the code. Named 'CODE_OVERVIEW.md' to fit neatly with the other 'C___.md' files.

callmetango commented on CODE_OVERVIEW.md:
```diff
@@ -0,0 +1,299 @@
+- **Resources**: Read `doc/` output. Read Xorg docs: freedesktop.org/wiki/Xorg.
+  Check GitHub issues and discussions.
+
+- **Contributing**: Use C99, follow warnings (strict flags).
```
I think it was best to move the paragraph to the `CONTRIBUTING.md` file and only add a link to the `CONTRIBUTING.md` file to avoid duplication.

callmetango commented on CODE_OVERVIEW.md:
```diff
@@ -0,0 +1,299 @@
+  Clean up and improve code. Optimize. Add features. Propose extensions?
+  Port to new platforms. Debug. Close issues. Write documentation.
+
+- **Community Interaction**: Join the Telegram/Matrix chat. Post in the
```
Same here: I think it was best to move the paragraph to the `README.md` file and only add a link to the `README.md` file to avoid duplication.

callmetango commented on CODE_OVERVIEW.md:
```diff
@@ -0,0 +1,299 @@
+
+#### Key Principles of the Codebase
+
+- **Modularity**: Code is split into libraries (e.g., `libxserver_*`) that are
```
I know this might be nitpicky but I would like to suggest to remove the bullet list and keep the paragraphs only. Since the keywords are already printed in bold, the meaning would remain the same, but readability would be somewhat facilitated. Normally we either use list items without blank lines inbetween or paragraphs.

callmetango commented on CODE_OVERVIEW.md:
```diff
@@ -0,0 +1,299 @@
+The Meson build system orchestrates compilation, handling dependencies,
+conditional builds, and configuration options.
+
+#### Key Principles of the Codebase
```
This should be a heading 2 = `##` according to the document structure.

the-real-dill replied:
Fixed.

callmetango reviewed (changes requested):
Thank you for your contribution! This looks really good and adds quite a lot of value. Please see my inline comments for things which may be improved a little bit.

the-real-dill commented on CODE_OVERVIEW.md:
```diff
@@ -0,0 +1,299 @@
+- **Important Headers**: `include/dix-config.h` (config macros),
+  `dix/dispatch.h` (request handlers), `os/osdep.h` (OS funcs).
+
+- **Debugging**: Use `-verbose` server flag. Enable libunwind for
```
I've incorporated these suggestions. Let me know if I got the wording right or if any further changes should be made.

callmetango commented on README.md:
```diff
@@ -117,7 +117,10 @@ Together we'll make X great again!
 ## Contact
 
-[XLibre Discussions at Github](https://github.com/orgs/X11Libre/discussions) | [XLibre mailing list at FreeLists](https://www.freelists.org/list/xlibre) | [@x11dev channel at Telegram](https://t.me/x11dev) | [#xlibre room at Matrix](https://matrix.to/#/#xlibre:matrix.org) | [XLibre security contact at Github](https://github.com/X11Libre/xserver/security/policy)
+- Post in the [XLibre GitHub discussions](https://github.com/orgs/X11Libre/discussions).
```
Could you please revert this change of the `README.md`? The contact area was deliberately designed to be very compact. We will also be adding further contact options via social media and therefore cannot “inflate” this area.

the-real-dill replied:
Not a problem, reverted. As with any of these changes, just let me know what should be adjusted.

callmetango commented on CODE_OVERVIEW.md:
```diff
@@ -0,0 +1,314 @@
+
+### Other Files
+
+- **Meson.build** and **meson_options.txt** provide build process and
```
> `Meson.build`

I know, nitpicking, but living a case sensitive live for decades this should be `meson.build`. 😉

callmetango reviewed (changes requested):
I think that's a very good start. LGTM except for the one very small change.

the-real-dill committed:
doc: add CODE_OVERVIEW.md

Add a new document providing an overview of the XLibre codebase.
Includes a high-level diagram of components, a summary of the
directory layout, and tips for exploring and working on the code.

Signed-off-by: 'The Real Dill' <the_real_dill_2000@proton.me>

metux commented:
@callmetango

metux commented:
Great work. Thanks.

Would you like to take care of more of the (in-code) documentation ? That would be a great help :)
